libsql-server/src/schema/scheduler.rs

```diff
@@ -1,70 +1,104 @@
 
-enum WorkResult {
-    Task {
-        old_status: MigrationTaskStatus,
-        task: MigrationTask,
-        error: Option<String>,
-    },
-    Schema {
-        job_id: i64,
-    },
-}
-
 pub struct Scheduler {
     namespace_store: NamespaceStore,
+    /// this is a connection to the meta store db, but it's used for migration operations
+    migration_db: Arc<Mutex<MetaStoreConnection>>,
     workers: JoinSet<WorkResult>,
+    /// A batch of tasks for the current job (if any)
     current_batch: Vec<MigrationTask>,
-    // job id of the curently processing job
+    /// Currently processing job
     current_job: Option<MigrationJob>,
     has_work: bool,
 }
 
 impl Scheduler {
+    pub(crate) fn new(
+        namespace_store: NamespaceStore,
+        mut conn: MetaStoreConnection,
+    ) -> crate::Result<Self> {
+        setup_schema(&mut conn)?;
+        Ok(Self {
+            namespace_store,
+            workers: Default::default(),
+            current_batch: Vec::new(),
+            current_job: None,
+            // initialized to true to kickoff the queue
+            has_work: true,
+            migration_db: Arc::new(Mutex::new(conn)),
+        })
+    }
+
     pub async fn run(mut self, mut receiver: mpsc::Receiver<SchedulerMessage>) {
-        let tasks_permits = Arc::new(Semaphore::new(10));
+        const MAX_CONCCURENT_JOBS: usize = 10;
+        let tasks_permits = Arc::new(Semaphore::new(MAX_CONCCURENT_JOBS));
         loop {
             tokio::select! {
                 Some(msg) = receiver.recv() => {
                     self.handle_msg(msg).await;
                 }
                 // There is work to do, and a worker slot to perform it
+                // TODO: optim: we could try enqueue more work in a go by try_acquiring more
+                // permits here
                 Ok(permit) = tasks_permits.clone().acquire_owned(), if self.has_work => {
                     self.enqueue_work(permit).await;
                 }
                 Some(res) = self.workers.join_next(), if !self.workers.is_empty() => {
-                    self.has_work = true;
                     match res {
+                        // TODO: handle schema update failure
                         Ok(WorkResult::Schema { job_id }) => {
                             let job = self.current_job.take().unwrap();
                             assert_eq!(job.job_id(), job_id);
-                            self.namespace_store.meta_store().update_job_status(job_id, MigrationJobStatus::RunSuccess).await.unwrap();
+                            with_conn_async(self.migration_db.clone(), move |conn| {
+                                update_job_status(conn, job_id, MigrationJobStatus::RunSuccess)
+                            })
+                            .await
+                                .unwrap();
+                            // the current job is finished, try to pop next one out of the queue:
+                            self.has_work = true;
```

Comment:
I thought that has_work == true means a job is not fininshed?

Reply:
has work means that there may be some work to do next, and we should try enqueueing some. When we finish the current job, there may be more work, we raise the flag to try to enqueue the next one.